uvtools/dspec.py

```diff
@@ -146,6 +147,250 @@ def delay_filter(data, wgts, bl_len, sdf, standoff=0., horizon=1., min_dly=0.0,
+                            cn_guess=None, use_linear=True, operator=None):
+    """
+    Fit a smooth model to 1D complex-valued data with flags, using a linear 
+    least-squares solver. The model is a Fourier series up to a specified 
+    order. As well as calculating a best-fit model, this will also return a 
+    copy of the data with flagged regions filled in ('in-painted') with the 
+    smooth solution. 
+    
+    Optionally, you can also add an uncorrelated noise realization on top of 
+    the smooth model in the flagged region.
+    
+    Parameters
+    ----------
+    data : array_like, complex
+        Complex visibility array as a function of frequency, with shape 
+        (Nfreqs,).
+    
+    flags : array_like, bool
+        Boolean flags with the same shape as data.
+    
+    sigma : float or array_like
+        Noise standard deviation, in the same units as the data. If float, 
+        assumed to be homogeneous in frequency. If array_like, must have 
+        the same shape as the data.
+        
+        Note that the choice of sigma will have some bearing on how sensitive 
+        the fits are to small-scale variations.
+    
+    nmax: int
+        Max. order of Fourier modes to fit. A model with complex Fourier modes 
+        between [-n, n] will be fitted to the data, where the Fourier basis 
+        functions are ~ exp(-i 2 pi n nu / (Delta nu).
+    
+    add_noise : bool, optional
+        Whether to add an unconstrained noise realization to the in-painted areas. 
+        This uses sigma to set the noise standard deviation. Default: False.
+    
+    cn_guess : array_like, optional
+        Initial guess for the series coefficients. If None, zeros will be used. 
+        A sensible choice of cn_guess can speed up the solver significantly. 
+        Default: None.
+    
+    use_linear : bool, optional
+        Whether to use a fast linear least-squares solver to fit the Fourier 
+        coefficients, or a slower generalized least-squares solver. 
+        Default: True.
+    
+    operator : array_like, optional
+        Fourier basis operator matrix. Must have shape (Nfreq, Nmodes), where 
+        Nmodes = 2*nmax + 1.
+    
+    Returns
+    -------
+    model : array_like
+        Best-fit model, composed of a sum of Fourier modes.
+        
+    model_coeffs : array_like
+        Coefficients of Fourier modes, ordered from modes [-n, +n].
+    
+    data_out : array_like
+        In-painted data.
+    """
+    # Construct Fourier basis operator if not specified
+    if operator is None:
+        # Frequency array
+        nu = np.arange(data.size)
+        L = nu[-1] - nu[0]
+        
+        # Build matrix operator for complex Fourier basis
+        n = np.arange(-nmax, nmax+1)
+        F = np.array([np.exp(-1.j * _n * nu / L) for _n in n])
+    else:
+        F = operator
+        n = np.arange(-nmax, nmax+1)
+        assert F.shape[0] == 2*nmax+1, "Fourier basis operator has the wrong shape"
```

Comment:
In this case, I feel more strongly that this should be in an actual `if...raise` block, since the user is explicitly providing their own operator that needs to be the right shape.

Reply:
Done!